server/Arcadia.Assistant.CSP/Vacations/CspEmployeeVacationsRegistry.cs

```diff
@@ -155,17 +196,133 @@ protected override void OnReceive(object message)
             }
         }
 
-        private async Task<IEnumerable<CalendarEventWithApprovals>> GetVacations()
+        private void FinishDatabaseRefresh(RefreshDatabase.Success msg)
         {
-            var vacations = await this.vacationsSyncExecutor.GetVacations(this.employeeId);
-            return vacations
-                .Where(v => this.IsCalendarEventActual(v.CalendarEvent))
+            this.UpdateDatabaseVacationsCache(msg.Events.ToList());
+            this.ScheduleNextDatabaseRefresh();
+        }
+
+        private void FinishGetCalendarEvents(GetCalendarEventsSuccess msg)
+        {
+            var vacations = msg.Events
+                .Where(e => this.IsCalendarEventActual(e.CalendarEvent))
+                .Select(x => x.CalendarEvent)
                 .ToList();
+            this.Sender.Tell(new GetCalendarEvents.Response(this.employeeId, vacations));
+
+            this.UpdateDatabaseVacationsCache(msg.Events.ToList());
+
+            this.ScheduleNextDatabaseRefresh();
         }
 
-        private Task<CalendarEventWithApprovals> GetVacation(string eventId)
+        private void UpdateDatabaseVacationsCache(List<CalendarEventWithApprovals> databaseVacations)
         {
-            return this.vacationsSyncExecutor.GetVacation(this.employeeId, eventId);
+            var createdEvents = new List<CalendarEventWithApprovals>();
```

Comment:
I'd have this logic in a separate class.
```
var newCache = new DatabaseVacations(list);
var diff = oldCache.diff(newCache)
foreach(diff.updatedEvents)
```

think OOP :P

Reply:
Yes, master...

Reply:
Done.